File: windows/ClipSync.Peer/Transport/ISyncTransport.cs
using System.Net.WebSockets;

namespace ClipSync.Peer.Transport;

/// <summary>One received WebSocket message, already assembled from fragments.</summary>
public abstract record TransportFrame
{
    private TransportFrame()
    {
    }

    public sealed record Text(string Payload) : TransportFrame;

    /// <summary>The peer closed the socket (or the socket died).</summary>
    public sealed record Closed : TransportFrame;

    /// <summary>A text message exceeded the configured limit; the engine must close.</summary>
    public sealed record TooLarge : TransportFrame;

    /// <summary>Protocol v1 forbids binary frames.</summary>
    public sealed record Binary : TransportFrame;
}

/// <summary>
/// Transport abstraction for one sync session so the engine stays independent of
/// server/client WebSocket wiring and tests can run in-memory.
/// </summary>
public interface ISyncTransport : IAsyncDisposable
{
    ValueTask<TransportFrame> ReceiveAsync(CancellationToken cancellationToken);

    ValueTask SendTextAsync(string payload, CancellationToken cancellationToken);

    ValueTask CloseAsync(WebSocketCloseStatus status, string description, CancellationToken cancellationToken);
}

File: android/app/src/main/java/com/clipsync/android/sync/ISyncTransport.kt
package com.clipsync.android.sync

/**
 * One received WebSocket message, already assembled from fragments.
 * Mirrors windows/ClipSync.Peer/Transport/ISyncTransport.cs so the session
 * engine can run against a fake in unit tests.
 */
sealed interface TransportFrame {
    data class Text(val payload: String) : TransportFrame

    data object Closed : TransportFrame

    data object TooLarge : TransportFrame

    data object Binary : TransportFrame
}

/**
 * Transport abstraction for one sync session. The engine never talks to OkHttp
 * directly; tests inject a fake.
 */
interface ISyncTransport {
    suspend fun receive(): TransportFrame

    suspend fun sendText(payload: String)

    suspend fun close(reason: String)
}
